tools/shield_utils.py

```diff
@@ -0,0 +1,368 @@
+# must be contained by an isolated NUMA node. Shielding this slice of the
+# NUMA node is done by moving all the existing tasks inside system cpuset,
+# letting the "shielded" and free for available for a new docker container.
+
+
+import os
+import re
+import io
+import sys
+import random
+import string
+# Add the file directory to PATH to be able to import other modules from within
+# the same directory.
+sys.path.append(os.path.dirname(__file__))
+import log_utils as log
+
+
+# Exit codes.
+EINVPATTERN = 1
+EINVSYSFS = 2
+ENOAVAILCPU = 3
+ECONFLICTCPUSET = 4
+EINVARG = 5
+ENOAVAILMEM = 6
+EOFFNODE = 7
+EDIRTYENV = 8
+ESHARENODE = 9
+ETEARDOWNSYS = 10
+EEXCEPT = 11
+ESKIP = 12
+
+
+# Patterns.
+CPUSET_MOUNT_PATTERN="^[^ ]+ (/.+) (?:cpuset |cgroup (?:[^ ]*,)?cpuset[, ])"
+SYSFS_MOUNT_PATTERN="sysfs (.+) sysfs .+"
+RANGE_PATTERN="([0-9][1-9]*)-([0-9][1-9]*)"
+
+
+def _locate_mount_point(pattern):
+    mount_point_regex = re.compile(pattern)
+    path = None
+    with open("/proc/mounts") as fp:
+        for line in fp:
+            res = mount_point_regex.search(line)
+            if res:
+                path = res.group(1)
+                break
+
+    return path
+
+
+# Parse list formats for cpuset and mems.
+# See: https://man7.org/linux/man-pages/man7/cpuset.7.html.
+#
+def _parse_list_format(content):
+    content = content.strip()
+    arr = set()
+    ranges=content.split(",")
+    for group in ranges:
+        m = re.search(RANGE_PATTERN, group)
+        # group is a singular value.
+        if m is None:
+            arr.add(int(group))
+        else: # group is a tuple.
+            group = (int(m.group(1)), int(m.group(2)))
+            arr.add(group)
+    return list(arr)
+
+
+def _is_node_online(numa_node_id):
+    sysfs_path = _locate_mount_point(SYSFS_MOUNT_PATTERN)
+    online_numa_nodes_path = sysfs_path + "/devices/system/node/online"
+    with open(online_numa_nodes_path) as fp:
+        online_numa_nodes = fp.readline().strip()
+        nodes = online_numa_nodes.split(",")
+        is_node_online = False
+        for node in nodes:
+            m = re.search(RANGE_PATTERN, node)
+            if m is None:
+                if int(node) == numa_node_id:
+                    return True
+
+            else:
+                range = (m.group(1), m.group(2))
+                if numa_node_id in range(range[0], range[1] + 1):
+                    return True
+    return False
+
+
+def _get_cpulist(numa_node_id, raw=False):
+    sysfs_path = _locate_mount_point(SYSFS_MOUNT_PATTERN)
+    cpulist_path = sysfs_path + "/devices/system/node/node{}/cpulist".format(numa_node_id)
+    with open(cpulist_path) as fp:
+        if raw:
+            return fp.readline()
+        return _parse_list_format(fp.readline())
+
+
+def _is_memory_node_available(numa_node_id):
+    cpuset_path = _locate_mount_point(CPUSET_MOUNT_PATTERN)
+    (_, dirnames, _) = next(os.walk(cpuset_path))
+    for dirname in dirnames:
+        if dirname != "docker":
```

Comment:
Removes an indentation block:
```suggestion
        if dirname == "docker":
            continue
```